### ISSUE TITLE
nix-build only autocalls functions with default values one level deep

### ISSUE THREAD

**Issue by ledettwy** [NixOS/nix#1978]
The following works when built with nix-build:
`with import <nixpkgs> {}; { arg2 ? null }: runCommand "test" {} ''touch $out;''`
But this does not:
`with import <nixpkgs> {}; { arg1 ? null }: { arg2 ? null }: runCommand "test" {} ''touch $out;''`

Seems like autocalling functions with default values only works one-level deep.

I'm using Nix 2.0

**Comment by peti** [comment]
Yes, this is true. I don't think that Nix is intended to behave any other way and I am not aware of documentation that would suggest that the use-case you gave would be supported.

I believe you should merge those two arguments into one attribute set argument if you want Nix to convert the lambda function into an attribute set automatically. If you have a function that returns yet another function, then you cannot use that feature.

**Comment by peti** [comment]
I'm closing this ticket in the light of the discussion that has already taken place at https://github.com/NixOS/nix/issues/1979.

**Comment by ledettwy** [comment]
> I'm closing this ticket in the light of the discussion that has already taken place at #1979.

@peti wait, why? #1979 is a different issue. I don't think there's a good reason to close this one yet.

> I believe you should merge those two arguments into one attribute set argument

Ah, this is why one might want a feature like this. Nix does not provide that many facilities for manipulating lambdas, so what you've suggested is easier said than done sometimes.

> I don't think that Nix is intended to behave any other way and I am not aware of documentation that would suggest that the use-case you gave would be supported.

From the nix-build man page entry for `--arg`:

> When evaluating Nix expressions, the expression evaluator will automatically try to call functions that it encounters. It can automatically call functions for which every argument has a default value (e.g., { argName ? defaultValue }: ...)

There's no implication or suggestion that this only works one level deep. Regardless of documentation though, there is value in nix-build doing this, and I don't see any value in limiting it to just one level.

Can you please reopen this?

**Comment by peti** [comment]
Well, ok. We can keep the ticket open if you'd like.

> There's no implication or suggestion that this only works one level deep.

The way I read that paragraph, it's pretty clear that this works only for the upper most level. Otherwise, the text would have to state explicitly that recursion takes place. Just my 2 cents.

**Comment by ledettwy** [comment]
@peti sure, at worst this is an enhancement request :)

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
